### DISCUSSION TITLE
Required props for Astro components

### DISCUSSION THREAD

**Discussion by Rolanddoda** [withastro/roadmap#176] [Proposal]
I am seeing this example in the docs:

```js
export interface Props {
  name: string;
  greeting?: string;
}

const { greeting = "Hello", name } = Astro.props as Props;
```

And tested it by NOT passing a name attribute but I don't get any kind of error or warning.

It would be a useful feature to have required props and get errors when props don't match the specified type.

----------------------
Also, the syntax for defining props looks a bit unintuitive to me:

```js
const { greeting = "Hello", name } = Astro.props
```

Would you consider having a different syntax? I believe a function-based one like:

```js
Astro.defineProps({
  name: { type: String, required: true },
  greeting: String
})
```

or with typescript:

```ts
interface Props {
  name: string
  greeting?: number
}

Astro.defineProps<Props>()
```

Having a function-based syntax will open doors for many things like specifying the type of a prop, validations of props, and whatnot. 

*Function-based syntax for props is inspired by Vue.js [defineProps](https://vuejs.org/api/sfc-script-setup.html#typescript-only-features)*

**Comment by sarah11918** [comment]
Hi @Rolanddoda ! This is the kind of question we would normally answer in our Discord at https://astro.build/chat

If you are able to join us there, then you can post questions like this in our support-thread channels, or chat with people in our general thread and you'll get much quicker answers to your questions!  This channel is not for questions about how to use Astro, but more for about designing Astro features.  :)

**Reply by Rolanddoda** [reply]
@sarah11918 in the docs I don't see anything about required props so this can be a feature request. I'll modify my post.

**Comment by MedLeon** [comment]
👍🏼 for a type-safe, validate-able solution.
